rust/src/types/amount.rs

```diff
@@ -0,0 +1,581 @@
+        **self == **other
+    }
+}
+
+impl Eq for u256 {}
+
+impl PartialOrd for u256 {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for u256 {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        let lhs: U256 = (*self).into();
+        let rhs: U256 = (*other).into();
+        lhs.cmp(&rhs)
+    }
+}
+
+impl u256 {
+    pub fn slt(&self, rhs: &Self) -> bool {
+        let lhs: I256 = (*self).into();
+        let rhs: I256 = (*rhs).into();
+        lhs.cmp(&rhs) == Ordering::Less
+    }
+
+    pub fn sgt(&self, rhs: &Self) -> bool {
+        let lhs: I256 = (*self).into();
+        let rhs: I256 = (*rhs).into();
+        lhs.cmp(&rhs) == Ordering::Greater
+    }
+}
+
+impl BitAnd for u256 {
+    type Output = Self;
+
+    fn bitand(mut self, rhs: Self) -> Self::Output {
+        for bit in 0..32 {
+            self[bit] &= rhs[bit];
+        }
+        self
+    }
+}
+
+impl BitOr for u256 {
+    type Output = Self;
+
+    fn bitor(mut self, rhs: Self) -> Self::Output {
+        for bit in 0..32 {
+            self[bit] |= rhs[bit];
+        }
+        self
+    }
+}
+
+impl BitXor for u256 {
+    type Output = Self;
+
+    fn bitxor(mut self, rhs: Self) -> Self::Output {
+        for bit in 0..32 {
+            self[bit] ^= rhs[bit];
+        }
+        self
+    }
+}
+
+impl Not for u256 {
+    type Output = Self;
+
+    fn not(mut self) -> Self::Output {
+        for bit in 0..32 {
+            self[bit] = !self[bit];
+        }
+
+        self
+    }
+}
+
+impl u256 {
+    pub fn byte(&self, index: Self) -> Self {
+        if index >= 32u8.into() {
+            return u256::ZERO;
+        }
+        let idx = index[31];
+        self[idx as usize].into()
+    }
+}
+
+impl Shl for u256 {
+    type Output = Self;
+
+    fn shl(self, rhs: Self) -> Self::Output {
+        if rhs > u256::from(255u8) {
+            return u256::ZERO;
+        }
+        let value: U256 = self.into();
+        let shift = rhs[31] as u32;
+        (value.wrapping_shl(shift)).into()
+    }
+}
+
+impl Shr for u256 {
+    type Output = Self;
+
+    fn shr(self, rhs: Self) -> Self::Output {
+        if rhs > u256::from(255u8) {
+            return u256::ZERO;
+        }
+        let value: U256 = self.into();
+        let shift = rhs[31] as u32;
+        (value.wrapping_shr(shift)).into()
+    }
+}
+
+impl u256 {
+    pub fn sar(self, rhs: Self) -> Self {
+        let negative = self[0] & 0x80 > 0;
+        if rhs > u256::from(255u8) {
+            if negative {
+                return u256::MAX;
+            } else {
+                return u256::ZERO;
+            }
+        }
+        let value: U256 = self.into();
+        let shift = rhs[31] as u32;
+        let mut shr = value.wrapping_shr(shift);
+        if negative {
+            shr |= U256::MAX.wrapping_shl(255 - shift);
+        }
+        shr.into()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Herbert told me that I don't need 100% coverage. Especially the methods on `u256` get tested by CT anyways. So I just put some tests there. The other modules (e.g. stack, memory, ...) will be tested fully.

Reply:
Since the focus of your work is not to deliver a production ready EVM but an evaluation of the practicality of Rust to be integrated in our software stack, a lack of coverage is OK. However, if there are techniques like fixtures or table-driven tests in Rust they should at one or the other place be demonstrated.